### PR TITLE
Task 2 (Army)

### PR DESCRIPTION
**Pull request by DmitriyVolik**

### CODE REVIEW

File: WarOOP/Models/Army.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+public class Army : IArmy
+{
+    public List<Warrior> Units;
+
+    public bool HasUnits
+    {
+        get
+        {
+            if (!Units[0].IsAlive)
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs]: What is the reason for this check?

File: WarOOP/Models/Army.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    
+    public void AddUnits(string type, int count)
+    {
+        for (int i = 0; i < count; i++)
+        {
+            Units.Add(Warrior.CreateWarrior(type));
+        }
+    }
+
+    public Warrior GetUnit() => Units[0];
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs]: What does this method return if there is no units in army?

File: WarOOP/Models/Warrior.cs [hunk 3]
```diff
@@ -17,7 +17,28 @@ public Warrior()
         Attack = 5;
     }
 
-    private void GetDamageFrom(Warrior enemy) => CurrentHealth -= enemy.Attack;
+    protected virtual void GetDamageFrom(Warrior enemy)
+    {
+        if (IsAlive) CurrentHealth -= enemy.Attack;
```
Comment (elkavitaly) [inline WarOOP/Models/Warrior.cs]: Use curly braces even for one line statements

File: WarOOP/Models/Warrior.cs [hunk 4]
```diff
@@ -17,7 +17,28 @@ public Warrior()
...
+    }
 
-    public void AttackTo(Warrior enemy) => enemy.GetDamageFrom(this);
-}
+    public void AttackTo(Warrior enemy)
+    {
+        if (IsAlive) enemy.GetDamageFrom(this);
+    }
+
+    public static Warrior CreateWarrior(string type)
```
Comment (elkavitaly) [inline WarOOP/Models/Warrior.cs]: I don't like the approach with strings and switch cases. Think about making it more type safe

File: WarOOP/Models/Army.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public List<Warrior> Units;
+
+    public bool HasUnits
+    {
+        get
+        {
+            if (!Units[0].IsAlive)
+            {
+                Units.Remove(Units[0]);
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs]: Why does the getter remove something in collection?

File: WarOOP/Models/Army.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+using System.Collections;
+using System.Threading.Channels;
+using WarOOP.Abstractions;
+
+namespace WarOOP.Models;
+
+public class Army : IArmy
+{
+    public List<Warrior> Units;
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs]: Why is this property public? Any other class can reassign and change the collection

File: WarOOP/Models/Warrior.cs [hunk 7]
```diff
@@ -17,7 +17,28 @@ public Warrior()
...
+    {
+        switch (type)
+        {
+            case "Warrior":
+                return new Warrior();
+            case "Knight":
+                return new Knight();
+            case "Defender":
+                return new Defender();
```
Comment (elkavitaly) [inline WarOOP/Models/Warrior.cs]: How did you compile the code if there is no Defender class yet?

File: WarOOP.Tests/ArmyTests.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            new object[] { "Warrior", 5, "Knight", 10 },
+            new object[] { "Knight", 10, "Knight", 12 },
+            new object[] { "Knight", 5, "Warrior", 9 },
+            new object[] { "Warrior", 10, "Warrior", 11 }
+        };
+
+    [Theory]
+    [MemberData(nameof(TestData1))]
+    public void FirstWinSecond(string armyType1, int armyCount1, string armyType2, int armyCount2)
```
Comment (elkavitaly) [inline WarOOP.Tests/ArmyTests.cs]: We agreed on other naming convention for tests.
Your tests don't cover edge-case scenarios. What if there is no units in army? Or you try to create units of wrong type?

File: WarOOP/Models/Army.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    public void AddUnits(string type, int count)
+    {
+        for (int i = 0; i < count; i++)
+        {
+            Units.Add(Warrior.CreateWarrior(type));
+        }
+    }
+
+    public Warrior GetUnit() => Units[0];
+
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs:36]: remove extra line

**Review by elkavitaly — CHANGES_REQUESTED** [review]

File: WarOOP/Models/Army.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
+using System.Collections;
+using System.Threading.Channels;
+using WarOOP.Abstractions;
+
+namespace WarOOP.Models;
+
+public class Army : IArmy
+{
+    private List<Warrior> Units;
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs]: I would add _readonly_ modifier. I don't think that you will need to reassign the collection

File: WarOOP/Models/Army.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            Units.Add(Warrior.CreateWarrior(type));
+        }
+    }
+    
+    public bool HasUnits()
+    {
+        if (Units.Count > 0 && !Units[0].IsAlive)
+        {
+            Units.Remove(Units[0]);
```
Comment (elkavitaly) [inline WarOOP/Models/Army.cs]: You just moved the logic to separate method. But it still has the same meaning. And this doesn't change anything. And again _HasUnits_ method's name doesn't sound like it should delete something. Try to think about logic without deleting any elements in collection

File: WarOOP/Models/Battle.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            throw new NullReferenceException("Army can not be null");
+        }
+        
+        if (!army1.HasUnits() && !army2.HasUnits())
+        {
+            throw new Exception("Armies can not be empty");
+        }
+
+        if (!army1.HasUnits())
```
Comment (elkavitaly) [inline WarOOP/Models/Battle.cs]: I see duplicated checks for empty armies

File: WarOOP/Models/Warrior.cs [hunk 13]
```diff
@@ -17,7 +17,22 @@ public Warrior()
         Attack = 5;
     }
 
-    private void GetDamageFrom(Warrior enemy) => CurrentHealth -= enemy.Attack;
+    protected virtual void GetDamageFrom(Warrior enemy)
+    {
+        if (IsAlive) {CurrentHealth -= enemy.Attack;}
```
Comment (elkavitaly) [inline WarOOP/Models/Warrior.cs]: Adding curly braces also means that you need to format the code. Now it's not formatted

File: WarOOP/Models/Warrior.cs [hunk 14]
```diff
@@ -17,7 +17,22 @@ public Warrior()
...
+    }
 
-    public void AttackTo(Warrior enemy) => enemy.GetDamageFrom(this);
-}
+    public void AttackTo(Warrior enemy)
+    {
+        if (IsAlive) {enemy.GetDamageFrom(this);}
+    }
+
+    public static Warrior CreateWarrior(Type type)
```
Comment (elkavitaly) [inline WarOOP/Models/Warrior.cs]: You can use generic type parameter and constraints to create new instance based on the type

File: WarOOP.Tests/ArmyTests.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+            {
+                new object[] { typeof(Warrior), 15, typeof(Knight), 7, true },
+                new object[] { typeof(Knight), 10, typeof(Knight), 7, true },
+                new object[] { typeof(Warrior), 5, typeof(Knight), 10, false },
+                new object[] { typeof(Knight), 10, typeof(Knight), 12, false },
+            };
+
+    [Theory]
+    [MemberData(nameof(TestData))]
+    public void ArmyAndArmy_Fight_Correct(Type armyType1, int armyCount1, Type armyType2, int armyCount2, bool expected)
```
Comment (elkavitaly) [inline WarOOP.Tests/ArmyTests.cs]: The name of the method you test should be on the first place

File: WarOOP.Tests/ArmyTests.cs [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    public void Fight_EmptyArmies_Exception()
+    {
+        var army1 = new Army();
+        var army2 = new Army();
+        
+        Assert.Throws<Exception>(()=> Battle.Fight(army1, army2));
+    }
+    
+    [Fact]
+    public void Fight_EmptyArmy_Correct()
```
Comment (elkavitaly) [inline WarOOP.Tests/ArmyTests.cs]: What does _correct_ mean? You need to put here an expected result. In this test expected result is false. So, you can name it _ReturnsFalse_ or just _false_

**Review by elkavitaly — CHANGES_REQUESTED** [review]

File: WarOOP/Models/Battle.cs (hunk 12 above)
Comment (DmitriyVolik) [reply WarOOP/Models/Battle.cs]: This is necessary for the automatic victory of one of the armies if the opposite is empty

**Review by elkavitaly — APPROVED** [review]
